Add virtio-balloon support

teawater opened the pull request:
This PR adds new option balloon to memory config.
Set it to on will open the balloon function.

sboeuf commented:
Hi @teawater, thank you very much for that PR, but given that we will support `virtio-mem` soon (thanks to your other PR), I don't see the benefit of having `virtio-balloon` as part of `Cloud-Hypervisor`.
Is there a concrete use case of `virtio-balloon` that is not covered by `virtio-mem`?

sameo commented:
> Hi @teawater, thank you very much for that PR, but given that we will support `virtio-mem` soon (thanks to your other PR), I don't see the benefit of having `virtio-balloon` as part of `Cloud-Hypervisor`.

I'd say `virtio-ballon` is the best option until `virtio-mem` makes it way upstream and matures.

teawater commented:
> Hi @teawater, thank you very much for that PR, but given that we will support `virtio-mem` soon (thanks to your other PR), I don't see the benefit of having `virtio-balloon` as part of `Cloud-Hypervisor`.
> Is there a concrete use case of `virtio-balloon` that is not covered by `virtio-mem`?

Thanks!  I think it is a good question.

Virtio-mem is a good memory elasticity feature.
But It is not yet merged into the kernel upstream.  And even if it is merged,  I think it will take some time for it to reach the product level.  We hope more people can use this feature.  That's why we added this feature support to the clh.

Virtio-balloon has some defects.  But It's been in the kernel for a long time.  This makes it more product-level.  That's why we added this feature to clh.

Best,
Hui

sboeuf commented on vmm/src/config.rs:
```diff
@@ -278,6 +281,8 @@ impl MemoryConfig {
                 mergeable_str = &param[10..];
             } else if param.starts_with("hotplug_size=") {
                 hotplug_str = &param[13..]
+            } else if param.starts_with("balloon=") {
```
Here I have a genuine question, I'm not sure how balloon is supposed to be handled. Is this another way of doing memory hotplug? It's important to understand what's the expectation is to understand if it should be added the same way `virtio-mem` is being added.
/cc @rbradford @sameo

sameo replied:
It's not doing memory hotplug, but I think it makes sense to add it as part of the memory configuration (and not as a standalone `virtio-ballon` switch).

sboeuf commented on vmm/src/memory_manager.rs:
```diff
@@ -453,14 +464,33 @@ impl MemoryManager {
+        } else if desired_ram < self.current_ram {
+            self.balloon_resize(self.current_ram - desired_ram)?;
         }
+        Ok(false)
```
I'm really confused about the way balloon is supposed to be handled, and that means I'm not sure if we should use the balloon along with the memory hotplug by default.
Usually the way it works, the management layer such as Kata Containers will ask for the balloon to be extended, this shouldn't be up to the VMM to choose this kind of thing.
So I think we might need a dedicated command to resize the balloon only.
/cc @rbradford @sameo

sameo replied:
I agree we should not use ballooning for the `resize` operation on the VM as it's effectively not resizing the VM but releasing some pages to the host. The VM memory size is not changed.
As you suggested, ballooning requests should come from higher levels of the stack, through a dedicated API that we should provide.

teawater replied:
Could you please give me some advises about the control interface of the balloon?

sameo replied:
We'd need a `/api/v1/vm.balloon` command that would take a size argument to set the new balloon size:

```Patch
diff --git a/vmm/src/api/openapi/cloud-hypervisor.yaml b/vmm/src/api/openapi/cloud-hypervisor.yaml
index c16e05a5..80d25af3 100644
--- a/vmm/src/api/openapi/cloud-hypervisor.yaml
+++ b/vmm/src/api/openapi/cloud-hypervisor.yaml
@@ -219,6 +219,22 @@ paths:
         500:
           description: The new device could not be added to the VM instance.
 
+  /vm.balloon:
+    put:
+      summary: Resize the VM virtio balloon device
+      requestBody:
+        description: The target size for the balloon device
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/VmBalloonResize'
+        required: true
+      responses:
+        204:
+          description: The VM balloon device was successfully resized.
+        404:
+          description: The VM balloon device could not be resized.
+
 components:
   schemas:
 
@@ -542,3 +558,11 @@ components:
       properties:
         id:
           type: string
+
+    VmBalloonResize:
+      type: object
+      properties:
+        size:
+          description: requested ballon size in bytes
+          type: integer
+          format: int64
```

teawater replied:
Added.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+            // The head contains the request type which MUST be readable.
+            if avail_desc.is_write_only() {
+                error!("The head contains the request type is not right");
+                continue;
```
I checked other code that use the virtio.  Most of them drop wrong package, error! and keep handle the next package.
So I followed their approach.

teawater replied:
Thanks!  Fixed.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                                } else {
+                                    Ok(())
+                                }
+                            };
```
This result is not for current function.
It will be send back to resize.work caller.

sboeuf replied:
Oh ok!

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                        if let Err(e) = self.inflate_queue_evt.read() {
+                            error!("Failed to get inflate queue event: {:?}", e);
+                            break 'epoll;
+                        } else if let Err(e) = self.process_queue(ev_type) {
```
If you don't mind, I would like to keep it because this PR only implements the basic functions of the balloon.  There are other vq need to be handled by this function.

sboeuf replied:
Ok fair enough :)

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                        if let Err(e) = self.deflate_queue_evt.read() {
+                            error!("Failed to get deflate queue event: {:?}", e);
+                            break 'epoll;
+                        } else if let Err(e) = self.process_queue(ev_type) {
```
Same with inflate.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+};
+use vmm_sys_util::eventfd::EventFd;
+
+const QUEUE_SIZE: u16 = 128;
```
This VQ size is defined by guest kernel.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+            0
+        } else {
+            1
+        };
```
Done.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                        libc::MADV_DONTNEED
+                    } else {
+                        libc::MADV_WILLNEED
+                    };
```
Done.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                    } else {
+                        libc::MADV_WILLNEED
+                    };
+                    let res = unsafe {
```
Done.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                        if let Err(e) = self.resize.evt.read() {
+                            error!("Failed to get resize event: {:?}", e);
+                            break 'epoll;
+                        } else {
```
Done.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                        break 'epoll;
+                    }
+                    PAUSE_EVENT => {
+                        debug!("PAUSE_EVENT received, pausing virtio-pmem epoll loop");
```
Done.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+                        }
+                    }
+                    _ => {
+                        error!("Unknown event for virtio-mem");
```
Done.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+const PAGE_SHIFT: u32 = 12;
+
+// Size of a PFN in the balloon interface.
+const VIRTIO_BALLOON_PFN_SHIFT: u64 = 12;
```
Done.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,569 @@
+    ResizeTriggerFail(DeviceError),
+}
+
+// Got from qemu/include/standard-headers/linux/virtio_balloon.h
```
Done.

teawater commented on vmm/src/config.rs:
```diff
@@ -255,6 +255,8 @@ pub struct MemoryConfig {
     #[serde(default)]
     pub hotplug_size: Option<u64>,
+    #[serde(default)]
+    pub balloon: bool,
```
Done.

sboeuf commented:
@teawater any progress here?

teawater commented:
> @teawater any progress here?

Thanks for your remind.

I'm busy with some company work.
I will come back to work on this PR as soon as I have time.

Best,
Hui

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,573 @@
+// Get resize event.
+const RESIZE_EVENT: DeviceEventT = 0;
+// New descriptors are pending on the virtio queue.
+const INFLATE_QUEUE_AVAIL_EVENT: DeviceEventT = 1;
```
Updated.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,573 @@
+// New descriptors are pending on the virtio queue.
+const INFLATE_QUEUE_AVAIL_EVENT: DeviceEventT = 1;
+// New descriptors are pending on the virtio queue.
+const DEFLATE_QUEUE_AVAIL_EVENT: DeviceEventT = 2;
```
Updated.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,573 @@
+            rx: None,
+            evt: self.evt.try_clone().map_err(Error::EventFdTryCloneFail)?,
+        })
+    }
```
What about split VirtioBalloonResize to VirtioBalloonResizeControler and VirtioBalloonResizeReceiver.
VirtioBalloonResizeControler will control the size of virtio-balloon.
VirtioBalloonResizeReceiver will receive the size from VirtioBalloonResizeControler.

teawater replied:
Split VirtioBalloonResize to VirtioBalloonResize and VirtioBalloonResizeReceiver.
Changed try_clone to get_receiver.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,573 @@
+
+impl Balloon {
+    // Create a new virtio-balloon.
+    pub fn new() -> io::Result<Balloon> {
```
Fixed.

teawater commented on vmm/src/api/http_endpoint.rs:
```diff
@@ -76,6 +77,9 @@ pub enum HttpError {
     VmAddNet(ApiError),
+
+    /// Could access to balloon
+    VmmBalloon(ApiError),
```
Fixed.

teawater commented on vmm/src/api/http_endpoint.rs:
```diff
@@ -562,3 +566,42 @@ impl EndpointHandler for VmAddNet {
     }
 }
+
+// /api/v1/vm.add-net handler
```
Fixed.

teawater commented on vmm/src/api/mod.rs:
```diff
@@ -127,6 +127,9 @@ pub enum ApiError {
     VmAddNet(VmError),
+
+    /// The VM balloon could not be resized
+    VmBalloonResize(VmError),
```
I changed all of VmBalloon to VmBalloonResize.

teawater commented on vmm/src/api/mod.rs:
```diff
@@ -526,3 +532,21 @@ pub fn vm_add_net(
+pub fn vm_balloon_resize(
+    api_evt: EventFd,
+    api_sender: Sender<ApiRequest>,
+    resize_data: u64,
```
Fixed.

teawater commented on vmm/src/vm.rs:
```diff
@@ -775,6 +775,14 @@ impl Vm {
         }
     }
 
+    pub fn balloon_resize(&mut self, resize_date: u64) -> Result<()> {
```
Changed to balloon_data.

teawater commented on vmm/src/lib.rs:
```diff
@@ -462,6 +462,19 @@ impl Vmm {
         }
     }
 
+    fn vm_balloon_resize(&mut self, resize_date: u64) -> result::Result<(), VmError> {
```
Changed to balloon_data.

teawater commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    fn send(&self, r: Result<(), Error>) {
+        self.tx.send(r).unwrap();
```
Updated according to https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1082.

teawater commented on vmm/src/memory_manager.rs:
```diff
@@ -545,6 +556,19 @@ impl MemoryManager {
+                .unwrap()
+                .resize(size)
+                .map_err(Error::VirtioBalloonResizeFail)?;
+            self.balloon_size = size;
```
Removed.

teawater commented on tests/integration.rs:
```diff
@@ -499,6 +499,15 @@ mod tests {
+
+        cmd.arg(format!("{}", resize));
+
+        cmd.status().expect("Failed to launch ch-remote").success()
```
Added.

teawater commented on tests/integration.rs:
```diff
@@ -3501,6 +3510,64 @@ mod tests {
+
+            // Add RAM to the VM
+            let desired_ram = 2048 << 20;
+            resize_command(&api_socket, None, Some(desired_ram));
```
Updated.

teawater commented:
@sboeuf @sameo I pushed a new version according to your comments.

sboeuf commented:
@teawater 
>@sboeuf @sameo I pushed a new version according to your comments.

Great, I'll review it on Monday!

sboeuf commented:
@sameo @rbradford PTAL

rbradford commented on vmm/src/api/http_endpoint.rs:
```diff
@@ -646,3 +647,42 @@ impl EndpointHandler for VmAddVsock {
+                match &req.body {
+                    Some(body) => {
+                        // Deserialize into a VmConfig
+                        let resize_data: u64 = match serde_json::from_slice(body.raw())
```
Why are you using JSON encoding APIs (here and in the remote) if you're only sending a single u64. Please define a struct with a single member to handle this.

rbradford commented:
Why can't we reuse the resize API, either by interpreting a request to shrink the memory as a request to increase the size of the balloo. Or by the addition of a "balloon" field.

teawater commented:
> Why can't we reuse the resize API, either by interpreting a request to shrink the memory as a request to increase the size of the balloo. Or by the addition of a "balloon" field.

@rbradford https://github.com/cloud-hypervisor/cloud-hypervisor/pull/947#discussion_r397806022
There are the comments around the question.

teawater commented:
Ω

> Why can't we reuse the resize API, either by interpreting a request to shrink the memory as a request to increase the size of the balloo. Or by the addition of a "balloon" field.

You mean add a endpoint /vm.balloon/resize ?

rbradford commented:
> > Why can't we reuse the resize API, either by interpreting a request to shrink the memory as a request to increase the size of the balloo. Or by the addition of a "balloon" field.
> 
> You mean add a endpoint /vm.balloon/resize ?

Add a "balloon" field to the `vmm::api::VmResizeData` struct. Use the existing `ch-remote resize` command and the existing vm.resize API.

rbradford commented:
@teawater any update on this?

teawater commented:
> @teawater any update on this?

I am sorry that my mailbox lost the letter about your review.  So I didn't get any update about this PR.
I will work on this PR this week.

Best,
Hui

teawater commented:
@rbradford I pushed a new version according to your comments.  Please help me review it.

Best,
Hui

rbradford commented on tests/integration.rs:
```diff
@@ -3743,6 +3748,52 @@ mod tests {
+                .args(&["--memory", "size=1024M,balloon=on"])
+                .args(&["--kernel", kernel_path.to_str().unwrap()])
+                .args(&["--cmdline", "root=PARTUUID=6fb4d1a8-6c8c-4dd7-9f7c-1fe0b9f2574c console=tty0 console=ttyS0,115200n8 console=hvc0 quiet init=/usr/lib/systemd/systemd-bootchart initcall_debug tsc=reliable no_timer_check noreplace-smp cryptomgr.notests rootfstype=ext4,btrfs,xfs kvm-intel.nested=1 rw"])
+                .default_disks()
```
Do you need to use custom kernel command line? Otherwise please either use stock kernel command line of use a firmware boot.

teawater replied:
Done.

rbradford commented on tests/integration.rs:
```diff
@@ -3743,6 +3748,52 @@ mod tests {
+
+            // Add RAM to the VM
+            let desired_ram = 1024;
+            resize_command(&api_socket, None, None, Some(desired_ram));
```
You don't test a reboot. What happens after a reboot? What should happen is that the balloon should remain the same size.

teawater replied:
Resize of balloon will not available after reboot.

rbradford replied:
The size of the balloon should be the same after a reboot as before.

rbradford commented on tests/integration.rs:
```diff
@@ -3743,6 +3748,52 @@ mod tests {
+            // Remove RAM to the VM
+            let desired_ram = 512;
+            resize_command(&api_socket, None, None, Some(desired_ram));
+
```
Can you test balloon in combination with memory hotplug?

teawater replied:
Done.

rbradford commented:
@teawater We updated our toolchain so please update your PR.

teawater commented:
Add new temp commit that update mem in BalloonEpollHandler after memory hotplug.
But it still cannot handle the issue that guest kernel reboot after memory hotplug and balloon deflate. :(

rbradford commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,664 @@
+            _ => return Err(Error::ProcessQueueWrongEvType(ev_type)),
+        };
+
+        error!("process_queue {:?}", self.mem);
```
Spurious debugging.

rbradford commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,664 @@
+    }
+
+    fn update_memory(&mut self, mem: &GuestMemoryMmap) -> std::result::Result<(), crate::Error> {
+        error!("update_memory {:?}", mem);
```
Spurious debugging.

rbradford commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,664 @@
+                            )));
+                        }
+                        // Fix me
+                        error!("update memory handler");
```
Spurious debugging.

sboeuf commented on vm-virtio/src/balloon.rs:
```diff
@@ -51,6 +51,8 @@ const DEFLATE_QUEUE_EVENT: DeviceEventT = 2;
 // The device should be paused.
 const PAUSE_EVENT: DeviceEventT = 4;
+// The device should update memory.
+const UPDATE_MEMORY_EVENT: DeviceEventT = 5;
```
Please remove this patch as this is not needed. The updated `GuestMemoryMmap` is contained inside the `GuestMemoryAtomic<GuestMemoryMmap>` object. This means it is up to date whenever the `virtio-balloon` device accesses the memory regions.

sboeuf commented on tests/integration.rs:
```diff
@@ -3648,6 +3661,22 @@ mod tests {
+            let desired_ram = 1024;
+            resize_command(&api_socket, None, None, Some(desired_ram));
+
+            thread::sleep(std::time::Duration::new(10, 0));
```
I started investigating this issue and I found out the memory hotplug was triggering a guest reboot as you mentioned.
I have tested with QEMU and I got the same issue. I'm wondering if the virtio-balloon driver supports memory hotplug.

sboeuf replied:
QEMU and Cloud-Hypervisor both work. About Cloud-Hypervisor, the fix has been included in this PR.

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -694,6 +699,15 @@ impl Vm {
+            self.memory_manager
+                .lock()
+                .unwrap()
+                .balloon_resize(desired_balloon << 20)
```
`desired_balloon` should already be the size in bytes, so please don't shift it here.

teawater replied:
Fixed.

sboeuf commented on vmm/src/memory_manager.rs:
```diff
@@ -798,6 +807,23 @@ impl MemoryManager {
         Ok(())
     }
 
+    pub fn balloon_resize(&mut self, size: u64) -> Result<(), Error> {
```
Replace `size` with `expected_ram`.

teawater replied:
Fixed.

sboeuf commented on src/bin/ch-remote.rs:
```diff
@@ -184,9 +186,16 @@ fn resize_api_command(
         None
     };
 
+    let desired_balloon: Option<u64> = if let Some(balloon) = balloon {
```
Sorry to be picky with the naming here, but `desired_balloon` sounds like the user is providing the size for the balloon, while actually it means the expected size of the RAM, by inflating the balloon.
So I suggest this variable should be called `desired_ram_w_balloon`. Please update everywhere this can apply.

teawater replied:
Fixed.

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -694,6 +699,15 @@ impl Vm {
+                .unwrap()
+                .balloon_resize(desired_balloon << 20)
+                .map_err(Error::MemoryManager)?;
+        }
```
You need to store the balloon size to restore it after a reboot. This should be part of a separate commit.
First thing you need is that `balloon_resize()` return the actual size of the balloon that has been calculated based on the `desired_ram_w_balloon`.
Second thing, you need to add a new `balloon_size: Option<u64>` as part of the `MemoryConfig` so that you can update it, and upon a reboot, the VMM should rely on this value to set the balloon.
/cc @rbradford

teawater replied:
Fixed.

sboeuf commented:
@davidhildenbrand just a quick question in case you might know. When running a VM with `virtio-balloon` enabled, and then hotplugging some new DIMMs into the VM, by inflating the balloon afterwards leads to a triple fault from the guest. Do you know if the `virtio-ballloon` driver is supposed to support such use case?
I tried with QEMU and it seems to fail too, but I'm not sure my testing is correct with QEMU.

davidhildenbrand commented:
It should work, although I wouldn't call it supported (there is just too much that can go wrong in the guest). Can you make sure that the hotplugged memory is actually getting onlined/used in the guest and you are not trying to inflate the balloon to sizes that the guest (not actually exposing hotplugged memory to the buddy) cannot handle?

Can you share
- /proc/meminfo before DIMM hotplug and after DIMM hotplug
- QEMU HMP/QMP "info memory_size_summary" before/after hotplug
- QEMU HMP/QMP "info balloon" before/after hotplug
- User kernel/QEMU version

That might help to understand what's happening.

Note that the "balloon" QEMU command always eats/reports the logical VM size, not the balloon size. That's a steady source of confusion and can result in errors after hotplugging memory (because that also makes the logical VM size change ...).

sboeuf commented:
Thanks for the detailed answer @davidhildenbrand 

>Can you make sure that the hotplugged memory is actually getting onlined/used in the guest and you are not trying to inflate the balloon to sizes that the guest (not actually exposing hotplugged memory to the buddy) cannot handle?

It's not onlined. I've tried using the `auto_online` feature from sysfs, but no luck there :(
That's not a big deal if I can't make it work with QEMU as long as you confirmed it's supposed to work. This means something's wrong with the virtio-balloon implementation in Cloud-Hypervisor then.

davidhildenbrand commented:
> Thanks for the detailed answer @davidhildenbrand
> 
> > Can you make sure that the hotplugged memory is actually getting onlined/used in the guest and you are not trying to inflate the balloon to sizes that the guest (not actually exposing hotplugged memory to the buddy) cannot handle?
> 
> It's not onlined. I've tried using the `auto_online` feature from sysfs, but no luck there :(
> That's not a big deal if I can't make it work with QEMU as long as you confirmed it's supposed to work. This means something's wrong with the virtio-balloon implementation in Cloud-Hypervisor then.

"not onlined", that's one of the things I referred to as "just too much that can go wrong in the guest". Which kernel are you on? See https://lkml.kernel.org/r/20200319131221.14044-1-david@redhat.com for more details regarding newer kernels. You have to set "/sys/devices/system/memory/auto_online_blocks" before hotplugging.

sboeuf commented on vm-virtio/src/balloon.rs:
```diff
@@ -304,14 +292,6 @@ impl BalloonEpollHandler {
-            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(UPDATE_MEMORY_EVENT)),
-        )
-        .map_err(DeviceError::EpollCtl)?;
-
```
@teawater This commit simply reverts the previous patch, but of course both commits must be removed before we can merge the PR. I just wanted to keep it this way so that you can take care of it.

sboeuf commented:
@teawater I fixed the issue you were seeing with memory hotplug here: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/947/commits/b519edef14b2ee65c9be4e12ff2f2491b4ec1cab
Please address the rest of the comments, but the PR is close from being ready now.

sboeuf commented:
@teawater ping!

teawater commented:
> @teawater ping!

Pong.  Please give me some time.

Best,
Hui

teawater commented:
@sboeuf @rbradford I pushed a new version according to your comnents.  Please help me review it.

rbradford commented on src/bin/ch-remote.rs:
```diff
@@ -516,6 +530,13 @@ fn main() {
+                .arg(
+                    Arg::with_name("balloon")
+                        .long("balloon")
+                        .help("New balloon size")
```
nit: Mention the units like memory.

teawater replied:
I changed this part to "help("New memory with balloon size")"

rbradford commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,605 @@
+
+            // The head contains the request type which MUST be readable.
+            if avail_desc.is_write_only() {
+                error!("The head contains the request type is not right");
```
This error message needs some work

teawater replied:
Added.

rbradford commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,605 @@
+                continue;
+            }
+            if avail_desc.len as usize % data_chunk_size != 0 {
+                error!("the request size {} is not right", avail_desc.len);
```
And this one.

teawater replied:
Added.

rbradford commented on vm-virtio/src/balloon.rs:
```diff
@@ -0,0 +1,605 @@
+            self.signal(&VirtioInterruptType::Queue, Some(&self.queues[queue_index]))?;
+        }
+
+        Ok(())
```
There are errors in this function that should be propagated up.

teawater replied:
All added.

rbradford commented on vmm/src/lib.rs:
```diff
@@ -489,9 +489,10 @@ impl Vmm {
         &mut self,
         desired_vcpus: Option<u8>,
         desired_ram: Option<u64>,
+        desired_ram_w_balloon: Option<u64>,
```
I don't really understand why this is like this when the balloon size is provided in the API.

teawater replied:
This is the memory size with balloon.  It should small than ram size.
Balloon can be used to decrease the memory usage of the guest.

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -743,6 +743,8 @@ impl Vm {
                 .unwrap()
                 .balloon_resize(desired_ram_w_balloon)
                 .map_err(Error::MemoryManager)?;
+            let balloon_size = self.config.lock().unwrap().memory.size - desired_ram_w_balloon;
```
Instead of calculating this a second time, could you get the balloon size from the `balloon_resize()` function? Make it return a `Result<u64>` instead of `Result<()>`.

teawater replied:
Updated

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -743,6 +743,8 @@ impl Vm {
                 .balloon_resize(desired_ram_w_balloon)
                 .map_err(Error::MemoryManager)?;
+            let balloon_size = self.config.lock().unwrap().memory.size - desired_ram_w_balloon;
+            self.config.lock().unwrap().memory.balloon_size = balloon_size;
```
Please add a comment on top of this, explaining that we update the configuration value for the balloon size to ensure a reboot would use the right value.

teawater replied:
Added.
